ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/PeriodicRecovery.java

```diff
@@ -186,31 +188,80 @@ public void shutdown (boolean async)
    }
 
     /**
-     * make all scanning operations suspend.
-     *
-     * This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
-     * threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
-     * progress when this method is called it will complete its scan without suspending.
-     *
-     * @param async false if the calling thread should wait for any in-progress scan to complete before returning
-     * @return the previous mode before attempting the suspend
+     * <p>
+     *     Make all scanning operations suspend.
+     * </p>
+     * <p>
+     *     This switches the recovery operation mode to SUSPENDED. Any attempt to start a new scan either by an ad hoc
+     *     threads or by the periodic recovery thread will suspend its thread until the mode changes. If a scan is in
+     *     progress when this method is called it will complete its scan without suspending.
+     * </p>
+     * <p>
+     *     Note that this method is also influenced by
+     *     <code>RecoveryEnvironmentBean.setWaitForFinalRecovery(boolean)</code>.
+     *     Please, refer to the its javadoc for further details.
+     *     In case <code>isWaitForFinalRecovery() == true</code>, it is <b>very important</b> that, before invoking
+     *     <code>suspendScan</code>:
+     *     * The Transaction System (i.e. TxControl) gets disabled (i.e. no new transactions will be created)
+     *     * The Transaction Reaper finishes monitoring all in-flight transactions
+     * </p>
+     * @param async false if the calling thread should wait for any in-progress scan to complete before returning.
+     *              In case <code>isWaitForFinalRecovery() == true</code>, this parameter is override.
+     * @return the previous mode before attempting the suspension
      */
 
    public Mode suspendScan (boolean async)
    {
        synchronized (_stateLock)
        {
-           // only switch and kick everyone if we are currently ENABLED
            Mode currentMode = getMode();
 
+           /*
+            * isWaitForFinalRecovery checks whether suspension should delay until RecoveryModules
+            * implementing SuspendBlockingRecoveryModule recover all their transactions.
+            */
+           if (currentMode == Mode.ENABLED && recoveryPropertyManager.getRecoveryEnvironmentBean().isWaitForFinalRecovery()) {
+
+               /*
```

Comment:
How would you feel about TransactionReaper.transactionReaper().waitForAllTxnsToTerminate(); being in here rather than needing to be called before suspending? We are just waiting IIUC so maybe delaying the wait till the point it is needed to be checked is OK?

Reply:
If we use `TransactionReaper.transactionReaper().waitForAllTxnsToTerminate()` here, we are going to exclude scenarios where:
* The Recovery Manager runs in a different server/JVM
* There are multiple Narayana/Arjuna instances running in different servers/JVMs